### PR TITLE
Adding pdf generation for Real World Ocaml

### PR DESCRIPTION
**Pull request by clecat**
> Disclaimer: Note that this work is still in progress and might encounter huge modifications, do not take the current state as permanent

This pull request adds a pdf generation for the book Real World Ocaml.
The generation is done in two parts:
- First, mdx is used to generate latex files from the markdown chapters.
- Then, book uses pdf latex generates a pdf with pdflatex, an hardcoded latex template and the generated latex files.

The code will need some cleaning, removing the useless parts and adding some features, but for now, the generation itself works pretty well.
I've encountered some bugs, with the generation of pdflatex files outside the build folder, but I'm investigating on that (it also has the side effect to break the table of content).

In order to generate the pdf, you simply have to execute the following command:
```
dune build @pdf
```
And the generated pdf should be there:
```
_build/default/static/book.pdf 
```

Be aware that you may need to install some pdflatex stuff in order to have this working.

This code uses this, not yet merged and prone to change, [pull request](https://github.com/realworldocaml/mdx/pull/188)

### CODE REVIEW

File: bin/lib/book.ml [hunk 1]
```diff
@@ -255,6 +255,22 @@ let make_simple_page file =
   ] in
   return (main_template ~title_bar:title_bar ~content ())
 
+let make_content_page ?(repo_root=".") () : string Deferred.t =
```
Comment (samoht) [inline bin/lib/book.ml]: maybe name this `make_tex_page` to be more explicit about what it does?

Comment (samoht) [reply bin/lib/book.ml]: Actually I am not sure to understand what `content.tex` is supposed to contain? Is that just the list of inputs? So maybe  name it `inputs.tex` instead?

File: duniverse/mdx/.travis.yml [hunk 2]
```diff
@@ -1,34 +1,19 @@
 language: c
+sudo: false
+services:
+  - docker
+install: wget https://raw.githubusercontent.com/ocaml/ocaml-ci-scripts/master/.travis-docker.sh
+script: bash ./.travis-docker.sh
```
Comment (samoht) [inline duniverse/mdx/.travis.yml]: why did you change that file?

File: duniverse/mdx/CHANGES.md [hunk 3]
```diff
@@ -2,30 +2,22 @@
 
 #### Added
 
-#### Changed
```
Comment (samoht) [inline duniverse/mdx/CHANGES.md]: why did you change that the beginning of that file?

File: duniverse/mdx/lib/block.ml [hunk 4]
```diff
@@ -15,6 +15,7 @@
  *)
 
 open Astring
+open Result
```
Comment (samoht) [inline duniverse/mdx/lib/block.ml]: You are not supposed to need this

File: duniverse/mdx/lib/library.ml [hunk 5]
```diff
@@ -1,3 +1,5 @@
+open Result
+
```
Comment (samoht) [inline duniverse/mdx/lib/library.ml]: same here, you are not supposed to need this

File: duniverse/mdx/lib/top/mdx_top.ml [hunk 6]
```diff
@@ -761,7 +761,7 @@ let rec save_summary acc s =
   | Env_value (summary, id, _) ->
      save_summary (Ident.name id :: acc) summary
 #if OCAML_MAJOR >= 4 && OCAML_MINOR >= 8
-  | Env_module (summary, id, _, _)
+  | Env_module (summary, id, Mp_present, _)
```
Comment (samoht) [inline duniverse/mdx/lib/top/mdx_top.ml]: Why is this needed?

File: duniverse/mdx/lib/library.ml [hunk 7]
```diff
@@ -40,7 +42,5 @@ module Set = struct
   end)
 
   let to_package_set t =
-    to_seq t
-    |> Seq.map (fun t -> t.base_name)
-    |> Astring.String.Set.of_seq
+    fold (fun t acc -> Astring.String.Set.add t.base_name acc) t Astring.String.Set.empty
```
Comment (samoht) [inline duniverse/mdx/lib/library.ml]: Why is this needed?

**Comment by samoht** [comment]
I've started commenting on the PR but actually it would be easier to review if you rebase it first over master (so that it would remove necessary changes). Let me know if you need some help to do this.

File: duniverse/mdx/.travis.yml (hunk 2 above)
Comment (clecat) [reply duniverse/mdx/.travis.yml]: This change and all the other comes are from the update of mdx, those are not done by me, but by updating mdx, I dragged them along

File: duniverse/mdx/CHANGES.md (hunk 3 above)
Comment (clecat) [reply duniverse/mdx/CHANGES.md]: Same as above

File: duniverse/mdx/lib/block.ml (hunk 4 above)
Comment (clecat) [reply duniverse/mdx/lib/block.ml]: Same as above

File: duniverse/mdx/lib/library.ml (hunk 7 above)
Comment (clecat) [reply duniverse/mdx/lib/library.ml]: Same as above

**Comment by clecat** [comment]
I'm currently doing a huge rework of the modifications, i'll try to drop the changes brought when rebasing on @NathanReb mdx duniverse branch

**Comment by NathanReb** [comment]
Actually, given that we plan on dropping `output`, I'd suggest you go ahead and import your work directly into RWO, using the public mdx library available in its duniverse.

See https://github.com/realworldocaml/book/issues/3213

**Comment by NathanReb** [comment]
It might be good to import mdx output in RWO first (as this is already part of the plan, see #3213)  and rebase this on top of it so that all the work needed for PDF generation sits in this PR!

**Comment by clecat** [comment]
There we go, I have done a huge rework on this PR:
It now depends on [this PR](https://github.com/realworldocaml/book/pull/3218) importing `ocaml-mdx output` from ocaml-mdx into book. This way we do not have to update the duniverse with more than 130 created/deleted/modified files.

The first commit adds the generation of Latex files in the new `rwo-output` binary. It's content comes from [this PR on mdx](https://github.com/realworldocaml/book/pull/3218) I closed in favor of this one.

The second one modifies `rwo-build` in order to create the needed files for the pdf generation and adds dune stanzas for the pdf creation.

Should still be done:
- Updating the `Dockerfile` file in order to pull the newly added ,dependencies (the texlive suite).
- Updating the documentation.

**Comment by samoht** [comment]
This needs to be rebased on master now that #3218 has been merged.

**Comment by clecat** [comment]
I'm on it, but I want to solve some small issues first, I'll push the merge as soon as possible

**Comment by samoht** [comment]
That looks good. Is `book.pdf` added to the html output so that it is available online?

**Comment by samoht** [comment]
Looks great, thanks for your work!
